core/embed/rust/src/ui/display/font.rs

```diff
@@ -106,7 +103,106 @@ impl Glyph {
 
-/// Font constants. Keep in sync with FONT_ definitions in
+/// A provider of font glyphs and their metadata.
+///
+/// Manages access to font resources and handles UTF-8 character glyphs
+///
+/// The provider holds necessary lock for accessing translation data
+/// and is typically used through the `Font::with_glyph_data` method
+/// to ensure proper resource cleanup.
+///
+/// # Example
+/// ```
+/// let font = Font::NORMAL;
+/// font.with_glyph_data(|data| {
+///     let glyph = data.get_glyph('A');
+///     // use glyph...
+/// });
+/// ```
+pub struct GlyphData {
+    font: Font,
+    #[cfg(feature = "translations")]
+    translations_guard: Option<RwLockReadGuard<'static, Option<Translations<'static>>>>,
+}
+
+impl GlyphData {
+    fn new(font: Font) -> Self {
+        #[cfg(feature = "translations")]
+        let translations_guard = flash::get().ok();
+
+        Self {
+            font,
+            #[cfg(feature = "translations")]
+            translations_guard,
+        }
+    }
+
+    pub fn get_glyph(&self, ch: char) -> Glyph<'_> {
+        let ch = match ch {
+            '\u{00a0}' => '\u{0020}',
+            c => c,
+        };
+        let gl_data = self.get_glyph_data(ch as u16);
+
+        ensure!(gl_data.is_some(), "Failed to load glyph");
+        Glyph::load(gl_data.unwrap())
+    }
+
+    fn calculate_glyph_size(header: &[u8]) -> usize {
+        let width = header[0] as i16;
+        let height = header[1] as i16;
+
+        let data_bytes = match constant::FONT_BPP {
+            1 => (width * height + 7) / 8, // packed bits
+            2 => (width * height + 3) / 4, // packed bits
+            4 => (width + 1) / 2 * height, // row aligned to bytes
+            8 => width * height,
+            _ => fatal_error!("Unsupported font bpp"),
+        };
+
+        5 + data_bytes as usize // header (5 bytes) + bitmap data
+    }
+
+    fn get_glyph_data(&self, codepoint: u16) -> Option<&[u8]> {
+        display::get_font_info(self.font.into()).map(|font_info| {
+            if codepoint >= ' ' as u16 && codepoint < 0x7F {
+                // ASCII character
+                unsafe {
+                    let ptr = *font_info
+                        .glyph_data
+                        .offset((codepoint - ' ' as u16) as isize);
+                    let header = slice::from_raw_parts(ptr, 2);
+                    let full_size = Self::calculate_glyph_size(header);
+                    slice::from_raw_parts(ptr, full_size)
+                }
+            } else {
+                #[cfg(feature = "translations")]
+                {
+                    if codepoint >= 0x7F {
+                        // UTF8 character from embedded blob
+                        if let Some(guard) = &self.translations_guard {
+                            if let Some(translations) = guard.as_ref() {
+                                if let Some(glyph) =
+                                    translations.get_utf8_glyph(codepoint, self.font as u16)
```

Comment:
how about something like
```suggestion
    if let Some(glyph) = &self
        .translations_guard
        .map(|g| g.as_ref())
        .map(|t| t.get_utf8_glyph(codepoint, self.font as u16)) {
```

Reply:
74a84516f9303f0b95cdd9a55e5457fb06898592